tests/parameter_flash_storage.cpp

```diff
@@ -0,0 +1,356 @@
+    mock("flash").expectOneCall("write"); // crc(data)
+    mock("flash").expectOneCall("lock");
+
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+
+    // We need to manually check expectations since we tuned the flash mock to preserve ordering.
+    mock("flash").checkExpectations();
+}
+
+static void err_cb(void *p, const char *id, const char *err)
+{
+    (void) p;
+    (void) id;
+    (void) err;
+    char msg[128];
+    snprintf(msg, sizeof msg, "MessagePack error on item \"%s\": \"%s\"", id, err);
+
+    FAIL(msg);
+}
+
+TEST(ConfigSaveTestCase, SavingConfigWorks)
+{
+    // Declare a parameter and gives it a value.
+    parameter_t p;
+    parameter_integer_declare(&p, &ns, "foo");
+    parameter_integer_set(&p, 10);
+
+    // Check that the flash writer is used
+    // Number of expected written bytes is implementation-dependent
+    // Change if if necessary
+    for (int i = 0; i < 8; i++) {
+        mock("flash").expectOneCall("write");
+    }
+
+    // Saves the parameter, then change its value
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+    parameter_integer_set(&p, 20);
+    CHECK_EQUAL(20, parameter_integer_get(parameter_find(&ns, "/foo")));
+
+    // Loads the config back from saved state
+    // We add an offset to skip the CRC and the block length
+    parameter_msgpack_read(&ns,
+                           (char *)(&data[PARAMETER_FLASH_STORAGE_HEADER_SIZE]),
+                           sizeof(data) - PARAMETER_FLASH_STORAGE_HEADER_SIZE,
+                           err_cb, NULL);
+
+    // Check that the parameter has the same value as saved
+    CHECK_EQUAL(10, parameter_integer_get(parameter_find(&ns, "/foo")));
+}
+
+TEST_GROUP(ConfigLoadTestCase)
+{
+    uint8_t data[128];
+    parameter_namespace_t ns;
+    parameter_t foo;
+
+    void setup()
+    {
+        mock("flash").ignoreOtherCalls();
+        parameter_namespace_declare(&ns, NULL, NULL);
+        parameter_integer_declare(&foo, &ns, "foo");
+    }
+};
+
+TEST(ConfigLoadTestCase, SimpleLoad)
+{
+    // Set a value, then save it
+    parameter_integer_set(&foo, 20);
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+
+    // Change the value
+    parameter_integer_set(&foo, 10);
+
+    // Load the tree
+    auto res = parameter_flash_storage_load(&ns, data);
+
+    // Value should be back to what it was
+    CHECK_TRUE(res);
+    CHECK_EQUAL(20, parameter_integer_get(&foo));
+}
+
+TEST(ConfigLoadTestCase, CRCIsChecked)
+{
+    // Set a value, then save it
+    parameter_integer_set(&foo, 20);
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+
+    // Change the value
+    parameter_integer_set(&foo, 10);
+
+    // Corrupt data
+    data[0] ^= 0x40;
+
+    // Load the tree
+    auto res = parameter_flash_storage_load(&ns, data);
+
+    // Value should not have changed
+    CHECK_FALSE(res);
+    CHECK_EQUAL(10, parameter_integer_get(&foo));
+}
+
+TEST(ConfigLoadTestCase, InvalidConfigReturnsFalse)
```

Comment:
[suggestion] Rename tests to avoid commenting them. For example this could be named `TEST(AParameterFlashStorageLoader, FailsGivenAnInvalidConfigCausedByParameterTreeLayoutChanges)`

Reply:
done.